File: 0378_Kth_Smallest_Element_in_a_Sorted_Matrix/solution.py
'''
378. Kth Smallest Element in a Sorted Matrix

Given a n x n matrix where each of the rows and columns are sorted in ascending order,
find the kth smallest element in the matrix.

Note that it is the kth smallest element in the sorted order, not the kth distinct element.

Example:

matrix = [
   [ 1,  5,  9],
   [10, 11, 13],
   [12, 13, 15]
],
k = 8,

return 13.

Reference
https://www.hrwhisper.me/leetcode-kth-smallest-element-sorted-matrix/
'''


class Solution(object):
    def kthSmallest_1(self, matrix, k):
        """
        :type matrix: List[List[int]]
        :type k: int
        :rtype: int
        """
        lo, hi = matrix[0][0], matrix[-1][-1]
        while lo <= hi:
            mid = lo + (hi - lo) >> 1
            temp = sum([self.binary_search(y, mid) for y in matrix])
            if temp >= k:
                hi = mid - 1
            else:
                lo = mid + 1
        return lo

    def binary_search(self, nums, x):
        low, high = 0, len(nums)
        while low < high:
            mid = (low + high) >> 1
            if nums[mid] <= x:
                low = mid + 1
            else:
                high = mid
        return low

    def kthSmallest_2(self, matrix, k):
        """
        :type matrix: List[List[int]]
        :type k: int
        :rtype: int
        """
        low, high = matrix[0][0], matrix[-1][-1]
        n = len(matrix)
        while low < high:
            mid = low + ((high - low) >> 1)
            temp = self.search_lower_than_mid(matrix, n, mid)
            if temp < k:
                low = mid + 1
            else:
                high = mid
        return low

    def search_lower_than_mid(self, matrix, n, x):
        i, j = n - 1, 0
        cnt = 0
        while i >= 0 and j < n:
            if matrix[i][j] <= x:
                j += 1
                cnt += i + 1
            else:
                i -= 1
        return cnt


matrix = [
    [1, 5, 9],
    [10, 11, 13],
    [12, 13, 15]
]

s = Solution()
print s.kthSmallest_1(matrix, 8)

File: 0378_Kth_Smallest_Element_in_a_Sorted_Matrix/solution.cpp
/*
378. Kth Smallest Element in a Sorted Matrix

Level: Medium

https://leetcode.com/problems/kth-smallest-element-in-a-sorted-matrix

# 240. Search a 2D Matrix II
*/

#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <numeric>
#include <algorithm>

#include "common/ListNode.h"
#include "common/TreeNode.h"

using namespace std;

/*
  Solution:  Binary Search

  Find the smallest x, such that there are k elements that are smaller or equal to x.

  Time complexity: O(nlogn*log(max – min))

  Space complexity: O(1)
*/

class Solution
{
public:
  int kthSmallest(vector<vector<int>> &matrix, int k)
  {
    const int n = matrix.size();
    long l = matrix[0][0];
    long r = matrix[n - 1][n - 1] + 1;
    while (l < r)
    {
      long m = l + (r - l) / 2;
      int total = 0;
      int s = n;
      for (const auto &row : matrix)
        total += (s = distance(begin(row), upper_bound(begin(row), begin(row) + s, m)));

      if (total >= k)
        r = m;
      else
        l = m + 1;
    }
    return l;
  }
};

class Solution
{
public:
  int kthSmallest(vector<vector<int>> &matrix, int k)
  {
    const int n = matrix.size();
    long l = matrix[0][0];
    long r = matrix[n - 1][n - 1] + 1;
    while (l < r)
    {
      // 注意优先级
      int mid = l + ((r - l) >> 1);
      int count = counter(matrix, mid);
      if (count < k)
        l = mid + 1;
      else
        r = mid;
    }
    return l;
  }

  /*
   is similar to 240. Search a 2D Matrix II. 
   The start point of such a sorted matrix is left-bottom corner.
  */
  int counter(vector<vector<int>> &matrix, int target)
  {
    int size = matrix.size();
    int row = size - 1;
    int col = 0;
    int count = 0;

    while (row >= 0 && col < size)
    {
      if (matrix[row][col] <= target)
      {
        count += row+1;  // add count in vertical
        col++;
      }
      else
      {
        row--;
      }
    }

    return count;
  }
};